enderio-base/src/main/java/com/enderio/base/client/ClientEventHandler.java

```diff
@@ -1,20 +1,78 @@
 package com.enderio.base.client;
 
+import com.enderio.base.common.handler.travel.TeleportHandler;
 import com.enderio.base.common.menu.SyncedMenu;
 import net.minecraft.client.Minecraft;
+import net.minecraft.client.particle.Particle;
+import net.minecraft.client.player.LocalPlayer;
+import net.minecraft.core.particles.ParticleTypes;
+import net.minecraft.util.Mth;
+import net.minecraft.world.phys.Vec3;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.event.TickEvent;
 import net.minecraftforge.eventbus.api.SubscribeEvent;
 import net.minecraftforge.fml.common.Mod;
+import org.apache.commons.lang3.tuple.Triple;
+
+import javax.annotation.Nullable;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.Optional;
 
 @Mod.EventBusSubscriber(value = Dist.CLIENT)
 public class ClientEventHandler {
+
+    private static int tick = 0;
+
     @SubscribeEvent
     public static void clientTick(TickEvent.ClientTickEvent e) {
-        if (e.phase == TickEvent.Phase.END && Minecraft.getInstance().player != null) {
-            if (Minecraft.getInstance().player.containerMenu instanceof SyncedMenu syncedMenu) {
+        LocalPlayer player = Minecraft.getInstance().player;
+        if (e.phase == TickEvent.Phase.END && player != null) {
+            tick++;
+            //Sync Menu
+            if (player.containerMenu instanceof SyncedMenu syncedMenu) {
                 syncedMenu.clientTick();
             }
+
+            //TravellingStaffParticle
+            if (player.isShiftKeyDown() && TeleportHandler.canTeleport(player) && tick%3==0) {
+                Optional<Vec3> pos = TeleportHandler.teleportPosition(player.getLevel(), player);
+                if (pos.isPresent()) {
+                    addParticle(pos.get());
+                }
+            }
+        }
+    }
+
+
+    private static void addParticle(Vec3 pos) {
```

Comment:
I've put it there, because I would need another TeleportHandler class just for the client stuff, because of classloading issues. Can extract that to another class called TeleportClientHandler but the code is really seperate anyway, so I don't see a big benefit for extracting it out

Reply:
The only value I see is that all the teleport functionality would then be in a single package. Someone looking at the teleport code for the first time might have a hard time finding this functionality.
Just a suggestion though, I am happy for it to stay where it is.

Reply:
moved it